activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -230,9 +258,11 @@ def within_new_transaction(options = {})
             yield
           rescue Exception => error
             if transaction
+              transaction.state.invalidate! if error.is_a? ActiveRecord::TransactionRollbackError
```

Comment:
Do deadlocks invalidate savepoints on all databases? Will this need to be available for adapters to control behavior?